### PR TITLE
feat: Add About activity and link it in Event Details

### LINKED ISSUE #119: Add about fragment and link in event details
An about fragment should be added just in case the about event info is more than 4 lines just as in the EventBrite app 

Working on it :+1:

### PR DESCRIPTION
**Pull request by dreadpool2**
Fixes #119 

videotogif_2018 06 13_18 53 57

### CODE REVIEW

**Comment by iamareebjamal** [comment]
Try reducing activities and use fragments, but alright. Take note from now on

**Comment by dreadpool2** [comment]
@iamareebjamal Please review this :+1:

**Comment by iamareebjamal** [comment]
Most of the stuff you have done programmatically in offset change listener can be done in XML

**Comment by dreadpool2** [comment]
@iamareebjamal Should I use DataBindingUtil?

**Comment by iamareebjamal** [comment]
We are not using databinding

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                detailsHeader.visibility = View.GONE
+                aboutEventCollapsingLayout.title = eventExtra.name
+                Timber.d("collapse 2"+eventExtra.name)
+                isHideToolbarView = !isHideToolbarView
+            }
+        } else if (percentage < 1f && !isHideToolbarView) {
+            //Not Collapsed
+            Timber.d("no collapse"+eventExtra.name)
+            detailsHeader.visibility = View.VISIBLE
+            aboutEventCollapsingLayout.title = " "
```
Comment (dreadpool2) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: @iamareebjamal I am facing problem in identifying the place where it can be used in xml actually.

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: *I am changing the visibility of the detailsHeader 
*Changing the title of collapsing layout 
These are all based on the toolbar view visibility

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: These all are based on collapsing toolbar's state AFAIK which can at least be used to change the title of itself

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: We can just set the way of viewing of collapsing toolbar on scrolling in xml, by attaching scroll flags, I couldn't find a way to set a title for the layout for each state in xml

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: Okay I think i found it :+1:

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: @iamareebjamal I have seemingly found a solution where you just set the title, and that offset listener is not even used. But like both the event name and date are on a same line

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: Actually just for the correct order of showing the details -> event name plus date, we need the offset listener. The title as you said will only have the event name, but we are setting the event name programitically itself

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: There is expandedTitle and collapsedTitle option in XML

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: That's what even I thought, but there isn't actually

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: https://stackoverflow.com/questions/31662416/show-collapsingtoolbarlayout-title-only-when-collapsed Second solution

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: Isn't that for just changing the text appearance ?

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: Okay I think I understood :+1:

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: That is a very buggy solution, it doesn't even work for some APIs and title still displays :(

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: It doesn't work with the latest support libraries

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: @iamareebjamal Programmitically is the best way to avoid bugs I think :+1:  What are your views ?

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: OK if you say so, but programmatic methods are the most buggy solutions. They are mostly used for non-standard solutions, can cause memory leaks and increase the development surface area, meaning we have more code to maintain, debug and fix. If there is no solution in your opinion, we may move forward

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -10,9 +10,10 @@
         android:icon="@mipmap/ic_launcher"
         android:label="@string/app_name"
         android:roundIcon="@mipmap/ic_launcher_round"
-        android:supportsRtl="true"
-        android:theme="@style/AppTheme">
-        <activity android:name=".MainActivity">
+        android:theme="@style/AppTheme"
+        android:supportsRtl="true">
+        <activity
+            android:name=".MainActivity">
```
Comment (iamareebjamal) [inline app/src/main/AndroidManifest.xml]: Avoid unneeded changes

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        val lowerCaseMonth = month.toString().toLowerCase()
+        val formatMonth = (lowerCaseMonth.substring(0, 1).toUpperCase() + lowerCaseMonth.substring(1)).substring(0, 3)
+
+        //Format Day
+        val day = dateString.month
+        val lowerCaseDay = day.toString().toLowerCase()
+        val formatDay = (lowerCaseDay.substring(0, 1).toUpperCase() + lowerCaseDay.substring(1)).substring(0, 3)
+
+        return formatMonth + ", " + formatDay + " " + dateString.dayOfMonth
+    }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: All logic inside ViewModel

**Comment by dreadpool2** [comment]
PR Updated :+1:

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+        if (supportActionBar != null) supportActionBar?.setDisplayHomeAsUpEnabled(true)
+        aboutEventCollapsingLayout.title = " "
+        appbar.addOnOffsetChangedListener(this)
+
+        aboutEventViewModel.error.observe(this, Observer {
+            Toast.makeText(this, it, Toast.LENGTH_LONG).show()
+        })
+
+        val intent = intent
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: Why?

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    private var isHideToolbarView: Boolean = false
+    private lateinit var eventExtra: Event
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.fragment_about_event)
+
+        setSupportActionBar(aboutEventToolbar)
+
+        if (supportActionBar != null) supportActionBar?.setDisplayHomeAsUpEnabled(true)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: `if (supportActionBar != null)` Not needed

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        if (percentage == 1f && isHideToolbarView) {
+            //Collapsed
+            if (TextUtils.isEmpty(eventExtra.name)) {
+                detailsHeader.visibility = View.GONE
+                aboutEventCollapsingLayout.title = eventExtra.name
+                isHideToolbarView = !isHideToolbarView
+            } else {
+                detailsHeader.visibility = View.GONE
+                aboutEventCollapsingLayout.title = " "
+                isHideToolbarView = !isHideToolbarView
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: When this is in both the cases, why not move it out of the if and else

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        val percentage =  Math.abs(verticalOffset).toFloat() / maxScroll.toFloat()
+
+        if (percentage == 1f && isHideToolbarView) {
+            //Collapsed
+            if (TextUtils.isEmpty(eventExtra.name)) {
+                detailsHeader.visibility = View.GONE
+                aboutEventCollapsingLayout.title = eventExtra.name
+                isHideToolbarView = !isHideToolbarView
+            } else {
+                detailsHeader.visibility = View.GONE
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: Same here

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt [hunk 8]
```diff
@@ -87,6 +88,20 @@ class EventDetailsFragment : Fragment() {
...
         }
 
+        //About event on-click
+        val aboutEventOnClickListener = View.OnClickListener {
+            val aboutIntent  = Intent(context, AboutEventActivity::class.java)
+            aboutIntent.putExtra(EVENT_ID, eventId)
+            startActivity(aboutIntent)
+        }
+
+        if(rootView.event_description.lineCount > 3){
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: Space around symbols

**Comment by dreadpool2** [comment]
Changed the top bar to white for better UI
screenshot_20180615-164413

**Comment by dreadpool2** [comment]
PR Updated :+1:

File: app/src/main/res/values/styles.xml [hunk 9]
```diff
@@ -11,9 +11,10 @@
     <style name="AppTheme.NoActionBar">
         <item name="windowActionBar">false</item>
         <item name="windowNoTitle">true</item>
+        <item name="colorPrimaryDark">@color/white</item>
```
Comment (nikit19) [inline app/src/main/res/values/styles.xml]: Can you make the white colour darker

Comment (dreadpool2) [reply app/src/main/res/values/styles.xml]: Do you mean grey ?

Comment (nikit19) [reply app/src/main/res/values/styles.xml]: Kind of. see the eventbrite app,the notifications of the user should be visible

Comment (dreadpool2) [reply app/src/main/res/values/styles.xml]: I think then the let the colour be kept as default colorPrimaryDark (blue)

**Comment by dreadpool2** [comment]
@iamareebjamal PR Updated

**Comment by dreadpool2** [comment]
@iamareebjamal Please review this :+1:

**Comment by iamareebjamal** [comment]
@nikit19 @simarsingh24 Please review

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        if (percentage < 1f && !isHideToolbarView) {
+            //Not Collapsed
+            detailsHeader.visibility = View.VISIBLE
+            aboutEventCollapsingLayout.title = " "
+            isHideToolbarView = !isHideToolbarView
+        }
+    }
+
+
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: extra line

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+                }, {
+                    error.value = "Error fetching event"
+                    Timber.e(it, "Error fetching event %d",id)
+                }))
+    }
+
+    fun getAboutEventFormattedDate(date: String): String{
+
+        val dateString = EventUtils.getLocalizedDateTime(date)
+
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt]: extra lines

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        val month = dateString.dayOfWeek
+        val lowerCaseMonth = month.toString().toLowerCase()
+        val formatMonth = (lowerCaseMonth.substring(0, 1).toUpperCase() + lowerCaseMonth.substring(1)).substring(0, 3)
+
+        //Format Day
+        val day = dateString.month
+        val lowerCaseDay = day.toString().toLowerCase()
+        val formatDay = (lowerCaseDay.substring(0, 1).toUpperCase() + lowerCaseDay.substring(1)).substring(0, 3)
+
+        return formatMonth + ", " + formatDay + " " + dateString.dayOfMonth
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt]: if timezone related information is also shown in Eventbrite app add that also here

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt [hunk 13]
```diff
@@ -89,6 +90,20 @@ class EventDetailsFragment : Fragment() {
...
         }
 
+        //About event on-click
+        val aboutEventOnClickListener = View.OnClickListener {
+            val aboutIntent  = Intent(context, AboutEventActivity::class.java)
+            aboutIntent.putExtra(EVENT_ID, eventId)
+            startActivity(aboutIntent)
+        }
+
+        if (rootView.event_description.lineCount > 3) {
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: declare 3 as a final const on top

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt (hunk 12 above)
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt]: Great will do that! :+1:

File: app/src/main/res/layout/content_event.xml [hunk 14]
```diff
@@ -129,20 +129,34 @@
...
+                android:layout_marginLeft="@dimen/layout_margin_large"
+                android:layout_marginRight="@dimen/layout_margin_large"
+                android:layout_height="wrap_content"
+                android:lines="1"
+                android:visibility="gone"
+                android:textStyle="bold"
+                android:layout_marginTop="4dp"
+                android:textSize="@dimen/text_size_small"
+                android:textColor="@color/black"
+                android:text="See more" />
```
Comment (nikit19) [inline app/src/main/res/layout/content_event.xml]: Please use a string resource

**Comment by dreadpool2** [comment]
@nikit19 @simarsingh24 @iamareebjamal Moved all possible dimensions to dimens.xml and strings to strings.xml and added time :+1:  Please Review
screenshot_20180617-010339

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt (hunk 10 above)
Comment (simarsingh24) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: still present

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+                    error.value = "Error fetching event"
+                    Timber.e(it, "Error fetching event %d",id)
+                }))
+    }
+
+    fun getAboutEventFormattedDate(date: String): String{
+        val dateString = EventUtils.getLocalizedDateTime(date)
+        //Format Month
+        val month = dateString.dayOfWeek
+        val lowerCaseMonth = month.toString().toLowerCase()
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt]: do `dateString.dayOfWeek.toString().toLowerCase()` directly

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+        val formatMonth = (lowerCaseMonth.substring(0, 1).toUpperCase() + lowerCaseMonth.substring(1)).substring(0, 3)
+        //Format Day
+        val day = dateString.month
+        val lowerCaseDay = day.toString().toLowerCase()
```
Comment (simarsingh24) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt]: same here

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt (hunk 12 above)
Comment (simarsingh24) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt]: did you add time zone info?

File: app/src/main/res/layout/content_event.xml [hunk 17]
```diff
@@ -128,39 +127,54 @@
...
+                android:textSize="@dimen/event_details_headers" />
 
             <TextView
                 android:id="@+id/eventDescription"
                 android:layout_width="match_parent"
                 android:layout_marginLeft="@dimen/layout_margin_large"
                 android:layout_marginRight="@dimen/layout_margin_large"
                 android:layout_height="wrap_content"
-                android:lines="3"
+                android:lines="4"
```
Comment (simarsingh24) [inline app/src/main/res/layout/content_event.xml]: why is lines 3 here and int LINE_COUNT=3 above?

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt (hunk 12 above)
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt]: Yup Z takes care about that

File: app/src/main/res/layout/content_event.xml (hunk 17 above)
Comment (dreadpool2) [reply app/src/main/res/layout/content_event.xml]: It is 4

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt (hunk 15 above)
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt]: Is that recommendable ?

File: app/src/main/res/layout/content_event.xml (hunk 17 above)
Comment (simarsingh24) [reply app/src/main/res/layout/content_event.xml]: why this is 4 here and LINE_count is 3 above?

Comment (dreadpool2) [reply app/src/main/res/layout/content_event.xml]: Actually the condition is more than 3 :+1:  not equal to 3

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt (hunk 15 above)
Comment (simarsingh24) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventViewModel.kt]: val month is not used anywhere else

File: app/src/main/res/layout/content_event.xml (hunk 17 above)
Comment (simarsingh24) [reply app/src/main/res/layout/content_event.xml]: okay

**Comment by dreadpool2** [comment]
PR Updated

File: app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+            it?.let {
+                loadEvent(it)
+            }
+        })
+    }
+
+    private fun loadEvent(event: Event){
+        eventExtra = event
+        aboutEventContent.text = event.description
+        val dateString = StringBuilder("")
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: No need to initialize StringBuilder with empty string

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: Did the changes

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: @nikit19 @simarsingh24 @iamareebjamal Please check whether any other modifications are required :+1:

Comment (nikit19) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: @dreadpool2 Can you rename the package to about. Since everything is about the events so aboutEvent does not convey much meaning according to me. It would also look similar to other packages like search,ticket etc

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/aboutEvent/AboutEventActivity.kt]: Seems legit :+1:  Any other changes required ?

File: app/src/main/res/layout/content_event.xml [hunk 19]
```diff
@@ -28,65 +28,64 @@
...
-                android:layout_marginBottom="10dp"
+                android:layout_marginBottom="@dimen/layout_margin_large"
                 android:orientation="vertical">
 
                 <TextView
                     android:id="@+id/eventName"
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
-                    android:text="Event Name"
+                    tools:text="Event Name"
```
[the remark this answers is not in the document]
Comment (dreadpool2) [reply app/src/main/res/layout/content_event.xml]: Did the changes :+1:

File: app/src/main/AndroidManifest.xml [hunk 20]
```diff
@@ -23,7 +23,11 @@
             android:name="org.fossasia.openevent.general.AuthActivity"
             android:windowSoftInputMode="adjustPan">
         </activity>
-        <activity android:name=".auth.LoginActivity"></activity>
```
Comment (iamareebjamal) [inline app/src/main/AndroidManifest.xml]: Why is this removed?

File: app/src/main/res/values/styles.xml [hunk 21]
```diff
@@ -13,7 +13,7 @@
         <item name="windowNoTitle">true</item>
     </style>
 
-    <style name="AppTheme.AppBarOverlay" parent="ThemeOverlay.AppCompat.Dark.ActionBar" />
+    <style name="AppTheme.AppBarOverlay" parent="ThemeOverlay.AppCompat.ActionBar"/>
```
[the remark this answers is not in the document]
Comment (iamareebjamal) [reply app/src/main/res/values/styles.xml]: So, it should only be used in new Activity and not globally

File: app/src/main/AndroidManifest.xml (hunk 20 above)
Comment (dreadpool2) [reply app/src/main/AndroidManifest.xml]: There doesn't exist any LoginActivity now

Comment (dreadpool2) [reply app/src/main/AndroidManifest.xml]: Its auth activity

File: app/src/main/res/values/styles.xml (hunk 21 above)
[the remark this answers is not in the document]
Comment (dreadpool2) [reply app/src/main/res/values/styles.xml]: @iamareebjamal I use this theme only in one file, should I still change it ?

File: app/src/main/AndroidManifest.xml (hunk 20 above)
Comment (iamareebjamal) [reply app/src/main/AndroidManifest.xml]: Still, these changes not related to issue should be handled in separate PR. Take note from next time on

File: app/src/main/res/values/styles.xml (hunk 21 above)
[the remark this answers is not in the document]
Comment (iamareebjamal) [reply app/src/main/res/values/styles.xml]: Name the style in a specific way

**Comment by dreadpool2** [comment]
@iamareebjamal @nikit19 @simarsingh24 Please CHeck :+1:
